crates/vcz_http_server/src/lib.rs

```diff
@@ -0,0 +1,654 @@
+    file_size: u64,
+    rh: http_range_header::ParsedRanges,
+) -> Result<Range<u64>, RhInFileError> {
+    match rh.validate(file_size) {
+        Err(e) => Err(RhInFileError::RangeUnsatisfiable(e)),
+        Ok(mut ranges) => if ranges.len() != 1 {
+            Err(RhInFileError::RangeNotOne)
+        } else {
+            Range::try_from_range(ranges.pop().unwrap())
+                .map_err(RhInFileError::RangeConvert)
+        },
+    }
+}
+
+
+
+fn decode_info_hash(info_hash: &str) -> Result<InfoHash, ()> {
+    let info_hash = info_hash.as_bytes();
+
+    /// The number of characters in a Base64 encoding of a SHA-1 hash.
+    const INFO_HASH_BASE64_CHAR_N: u8 = INFO_HASH_BIT_N
+        .div_ceil(6)
+        .div_ceil(4)
+        * 4;
+
+    if info_hash.len() != INFO_HASH_BASE64_CHAR_N.into() { Err(()) } else {
+        match URL_SAFE.decode(info_hash) {
+            Err(_) => Err(()),
+            Ok(info_hash) => info_hash.try_into().map_err(|_| ()),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum ParseUriPathError {
+    #[error("splitting the URI path yielded zero elements")]
+    Split,
+
+    #[error("the URI path should start with '/'")]
+    StartSlash,
+
+    #[error("the URI path should start with an info hash")]
+    NoInfoHash,
+
+    #[error("invalid info hash in the URI path")]
+    InvalidInfoHash,
+
+    #[error("a URI path component is not encoded with UTF-8")]
+    FromUtf8,
+}
+
+/// Web routing.
+/// Validates and converts a URI path to a file in a torrent.
+/// If it returns `Ok((info_hash, path))`, then `uri_path` refers
+/// to the file in the torrent identified by `info_hash`
+/// at `path` relative to the root of the torrent.
+fn parse_uri_path(
+    uri_path: &str,
+) -> Result<(InfoHash, Vec<String>), ParseUriPathError> {
+    let mut uri_path = uri_path.split('/').collect::<VecDeque<_>>();
+    match uri_path.pop_front() {
+        None => Err(ParseUriPathError::Split),
+        Some(first) => if !first.is_empty() {
+            Err(ParseUriPathError::StartSlash)
+        } else {
+            match uri_path.pop_front() {
+                None => Err(ParseUriPathError::NoInfoHash),
+                Some(info_hash) => match decode_info_hash(info_hash) {
+                    Err(_) => Err(ParseUriPathError::InvalidInfoHash),
+                    Ok(info_hash) => {
+                        let path: Result<Vec<_>, _> = Result::from_iter(
+                            uri_path
+                                .into_iter()
+                                .map(|s| urlencoding::decode(s)
+                                    .map(std::borrow::Cow::into_owned)
+                                )
+                        );
+                        match path {
+                            Err(_) => Err(ParseUriPathError::FromUtf8),
+                            Ok(path) => Ok((info_hash, path)),
+                        }
+                    },
+                },
+            }
+        },
+    }
```

Comment:
this code can be simplified:

```suggestion
    match uri_path.pop_front() {
        None => Err(ParseUriPathError::Split),
        Some(first) if !first.is_empty() => Err(ParseUriPathError::StartSlash),
        Some(first) => {
            let info_hash = uri_path.pop_front().ok_or(ParseUriPathError::NoInfoHash)?;
            let info_hash = decode_info_hash(info_hash)
                .map_err(ParseUriPathError::InvalidInfoHash)?;
            uri_path
	            .into_iter()
	            .map(|s| urlencoding::decode(s)
	                .map(std::borrow::Cow::into_owned)
	            )
	            .collect::<Result<Vec<_>, _>>()
	            .map_err(ParseUriPathError::FromUtf8)
        },
    }
```